File: Part 2 - Regression/Section 8 - Decision Tree Regression/decision_tree_regression_me.R
dataset=read.csv('Position_Salaries.csv')
dataset=dataset[2:3]

#Splitting the dataset into training and test set
#install.packages('caTools')
# 
# library(caTools)
# set.seed(123)
# split=sample.split(dataset$Purchased,SplitRatio = 0.8)
# training_set=subset(dataset,split==TRUE)
# test_set=subset(dataset,split==FALSE)

#Feature SCaling Excluding categorical values
# training_set[,2:3]=scale(training_set[,2:3])
# test_set[,2:3]=scale(test_set[,2:3])


#Fitting Decision Tree Regression Model to the dataset
# install.packages('rpart')
library(rpart)
regressor=rpart(formula = Salary ~ .,
                data = dataset,
                control = rpart.control(minsplit = 1))
#Normally split then whole thing is in one split.
#So the constant line comes.


#Predicting a new result with Decision Tree Regression Model
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))


# #Visualising the Decision Tree Regression Model results
# library(ggplot2)
# ggplot() +
#   geom_point(aes(x =dataset$Level , y =dataset$Salary),
#              colour = 'red') +
#   geom_line(aes(x =dataset$Level , y =predict(regressor, newdata = dataset)),
#             colour = 'blue') +
#   ggtitle('Truth or Bluff (Decision Tree Regression Model)') +
#   xlab('Level')+
#   ylab('Salary')


#Visualising the Decision Tree Regression Model results (Creating Smooth curve and increase the resolution)
library(ggplot2)
x_grid=seq(min(dataset$Level), max(dataset$Level),0.01)
ggplot() +
  geom_point(aes(x =dataset$Level , y =dataset$Salary),
             colour = 'red') +
  geom_line(aes(x =x_grid , y =predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Decision Tree Regression Model)') +
  xlab('Level')+
  ylab('Salary')

File: Part 2 - Regression/Section 8 - Decision Tree Regression/decision_tree_regression_me.py
# Decision Tree Regression Model is not a good model in two variables.
import numpy as np # Non Linear and Non Continuos Regression Model
import matplotlib.pyplot as plt
import pandas as pd
df=pd.read_csv('Position_Salaries.csv')
X=df.iloc[:,1:2].values # X is a matrix
y=df.iloc[:,-1].values # y is a vector

#Splitting the dataset into training and test set (Here we dont have many observation)
# and we have to make very accurate prediction. So in this case we are taking the whole dataset to train the model. 
"""from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)"""

#Feature Scaling, 
"""from sklearn.preprocessing import StandardScaler
sc_X=StandardScaler()
X_train=sc_X.fit_transform(X_train)
X_test=sc_X.transform(X_test) """

#Fitting Decision tree Model to the dataset 
from sklearn.tree import DecisionTreeRegressor
regressor=DecisionTreeRegressor(random_state=0)
regressor.fit(X,y)

#Prdicting a new result
y_pred=regressor.predict(6.5)

"""#Visualising Decision Tree results
plt.scatter(X,y,color='red')
plt.plot(X,regressor.predict(X),color='blue')
plt.title('Truth or Bluff (Regression Model)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()""" #As non continous and non Linear so
# it will give us fault.lecture 66.

#Visualising Regression results(for higher resolution and smoother curve)
X_grid=np.arange(min(X),max(X),0.01)
X_grid=X_grid.reshape((len(X_grid),1))
plt.scatter(X,y,color='red')
plt.plot(X_grid,regressor.predict(X_grid),color='blue')
plt.title('Truth or Bluff (Decision Tree Regression Model)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show() # Spliting is clearly seen
# In each of the splitting value is the mean (constant) as told in
# the intuition lecture.